### ISSUE TITLE
Potential precision issue of the fused linear operator

### ISSUE THREAD

**Issue by noc-turne** [BobMcDear/attorch#10]
Hi, thank you for your great work on this project!

I recently encountered a potential precision issue when calling the forward function of your fused MLP operator. For testing, I set the weight of the linear layer as 1, bias as 0 and the activation function as GELU. 

Below is my test code, simply place the test code at the same directory level as the `attorch` library:

```python
import torch
import torch.nn as nn
from attorch.attorch import Linear

class FusedMLP(nn.Module):
    def __init__(self, hidden_size, intermediate_size):
        super().__init__()
        self.act = nn.GELU()
        self.fc1 = nn.Linear(hidden_size, intermediate_size)

    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
        hidden_states = self.fc1(hidden_states)
        hidden_states = self.act(hidden_states)
        return hidden_states

def accuracy_test():
    hidden_size = 2
    intermediate_size = 4

    x = torch.rand(1, 2, device=torch.device('cuda'))
    print(x)
    torch_linear = FusedMLP(hidden_size, intermediate_size)
    attorch_linear = Linear(hidden_size, intermediate_size, act_func='gelu')

    with torch.no_grad():
        torch_linear.fc1.weight = nn.Parameter(torch.ones(intermediate_size, hidden_size))  # set to 1
        torch_linear.fc1.bias = nn.Parameter(torch.zeros(intermediate_size))  # set to 0

    torch_linear = torch_linear.to('cuda')

    with torch.no_grad():
        attorch_linear.weight = nn.Parameter(torch.ones(hidden_size, intermediate_size))  # set to 1
        attorch_linear.bias = nn.Parameter(torch.zeros(intermediate_size))  # set to 0

    attorch_linear = attorch_linear.to("cuda")

    y_torch = torch_linear(x)
    y_attorch = attorch_linear(x)
    print(torch.allclose(y_attorch, y_torch))
    print(y_attorch)
    print(y_torch)

accuracy_test()
```

A possible result is as follows:

```python
tensor([[0.9226, 0.2851]], device='cuda:0')
False
tensor([[1.0701, 1.0701, 1.0701, 1.0701]], device='cuda:0',
grad_fn=)
tensor([[1.0705, 1.0705, 1.0705, 1.0705]], device='cuda:0',
grad_fn=)
```

I have tested your activation function alone, and there is no precision problem. Would you mind taking a look to see if there's a potential bug in the linear function?

Thanks in advance for your time and help!

**Comment by BobMcDear** [comment]
Hello,

Small numerical discrepancies should be expected when comparing Triton and PyTorch, especially in artificial scenarios such as this one. In practice, this is almost never a problem, and the network will learn just fine. To convince yourself that this is the case, you can take a look at [the WikiText-2 example](https://github.com/BobMcDear/attorch/tree/main/examples/wikitext-2), where PyTorch and attorch (with fused MLPs) converge to the same loss. There are also [tests](https://github.com/BobMcDear/attorch?tab=readme-ov-file#tests) for each module to ensure correctness in a more rigorous fashion.

**Comment by noc-turne** [comment]
Okay, got it. Just out of curiosity, what causes this small numerical discrepancy?

**Comment by BobMcDear** [comment]
Floating-operations are by nature imprecise - in fact, they don't even conform to the usual rules of associativity. When computing the dot product of two vectors ```[a1, a2, a3, a4]``` and ```[b1, b2, b3, b4]```, this means the result depends on the order in which the product of corresponding elements are added; that is, ```(a1 * b1 + a2 * b2) + (a3 * b3 + a4 * b4)``` mightn't equal ```a1 * b1 + a2 * b2 + a3 * b3 + a4 * b4```, leading to small errors. Parallel code is especially susceptible to this problem since numerous blocks of values are processed separately and afterwards gathered together, often in an unpredictable sequence.

**Comment by noc-turne** [comment]
Thanks a lot for your time and help!
